### ISSUE TITLE
Support formatting jupyter notebooks

### ISSUE THREAD

**Issue by konstin** [astral-sh/ruff#1218]
[black](https://github.com/psf/black) supports formatting jupyter notebook with the jupyter extra `black[jupyter]`. If installed like this, it formats jupyter notebooks just like python files. It would be nice if ruff could similarly lint and fix jupyter notebooks.

Steps to implement this:

- [x] Read Jupyter notebooks, [build a mapping between cells and concatenated code](https://github.com/charliermarsh/ruff/blob/cab65b25da8aff6fb9cb539b55345aae64a835f0/crates/ruff/src/jupyter/notebook.rs#L136-L187) and and print lint errors with cell ids. The code is in [jupyter/notebooks.rs](https://github.com/charliermarsh/ruff/blob/main/crates/ruff/src/jupyter/notebook.rs).
- [x] Apply fixes to concatenated code and map back to cells
- [x] Write back jupyter notebooks ensuring that if there are no changes the json is identical (roundtrip support). You might need to extend or replace the current [schema](https://github.com/charliermarsh/ruff/blob/595cd065f3a002a4301c41f005122a8806a3925f/crates/ruff/src/jupyter/schema.rs#L23) or drop some of the schema for just `serde_json::Value`. See also the [black implementation](https://github.com/psf/black/blob/69ca0a4c7a365c5f5eea519a90980bab72cab764/src/black/__init__.py#L1017-L1046).
- [x] Test roundtrip support with a bunch of different notebooks found in the wild, ideally from different generators (e.g. jupyter notebook and pycharm) and different schema versions.
- [ ] Remove `jupyter_notebook` feature and include `.ipynb` files in ruff by default
- [ ] Optional: Check if there a any lints that should be off for jupyter notebook, e.g. "no print" or isort rules might not make sense

**Comment by charliermarsh** [comment]
I assume Ruff would work with [nbQA](https://github.com/nbQA-dev/nbQA), but I too would like it to support Jupyter out-of-the-box. Looking into it...

**Comment by charliermarsh** [comment]
Ruff was actually integrated into [nbQA](https://github.com/nbQA-dev/nbQA/pull/773) in the latest release, so you can now run (e.g.):

```
❯ nbqa ruff Untitled.ipynb
Untitled.ipynb:cell_1:2:5: F841 Local variable `x` is assigned to but never used
Untitled.ipynb:cell_2:1:1: E402 Module level import not at top of file
Untitled.ipynb:cell_2:1:8: F401 `os` imported but unused
Found 3 error(s).
1 potentially fixable with the --fix option.
```

(Would still like to have a first-party integration at some point.)

**Comment by blakeNaccarato** [comment]
See a parallel effort over in the Ruff VSCode extension, where LSP support for analyzing entire Jupyter notebooks seems blocked by pygls still being on LSP v3.16. It could be a shorter road for the Ruff LSP implementation to handle entire notebook formatting, if effort can be concentrated over at `pygls` to get LSP 3.17 out. See [my comment](https://github.com/charliermarsh/ruff-vscode/issues/51#issuecomment-1463190089) in the other thread for more detail and other Issue links.

It would be nice to have native support in Ruff as well! But in the short run, it looks like LSP support may have less resistance.

**Comment by sladyn98** [comment]
@konstin I could take this on

**Comment by charliermarsh** [comment]
@sladyn98 - Sorry for the churn, I already chatted with @dhruvmanila about taking this one on!

**Comment by dhruvmanila** [comment]
@charliermarsh you can assign this to me to avoid any future confusion :)

**Comment by dhruvmanila** [comment]
Sorry, not yet completed.

**Comment by charliermarsh** [comment]
What's the outstanding work here? Testing?

**Comment by dhruvmanila** [comment]
Meta issue: #5188
